hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/oauth2/CustomTokenProviderAdapter.java

```diff
@@ -46,16 +48,53 @@
    *
    * @param adaptee the custom token provider
    */
-  public CustomTokenProviderAdapter(CustomTokenProviderAdaptee adaptee) {
+  public CustomTokenProviderAdapter(CustomTokenProviderAdaptee adaptee, int customTokenFetchRetryCount) {
     Preconditions.checkNotNull(adaptee, "adaptee");
     this.adaptee = adaptee;
+    fetchTokenRetryCount = customTokenFetchRetryCount;
   }
 
   protected AzureADToken refreshToken() throws IOException {
     LOG.debug("AADToken: refreshing custom based token");
 
     AzureADToken azureADToken = new AzureADToken();
-    azureADToken.setAccessToken(adaptee.getAccessToken());
+
+    String accessToken = null;
+
+    Exception ex;
+    boolean succeeded = false;
+    // Custom token providers should have their own retry policies,
+    // Providing a linear retry option for the the retry count
+    // mentioned in config "fs.azure.custom.token.fetch.retry.count"
+    int retryCount = fetchTokenRetryCount;
+    do {
+      ex = null;
+      try {
+        accessToken = adaptee.getAccessToken();
+        LOG.trace("CustomTokenProvider Access token fetch was successful with retry count {}", retryCount);
```

Comment:
Is it necessary to add trace log here? it affects the perf for all rereshToken requests including those succeed without retry.
And the `retryCount` in the message is confusing,  the total times of retries that have been executed should be `fetchTokenRetryCount - retryCount`, same for the msg in the `catch` block.

Reply:
Retaining the trace log as we have had couple of debugging scenarios where customTokenProvider implementation hung on execution. Trace log right after the adapter.getAccessToken() will provide quick indication of the issue.
From previous PRs, have confirmation from Steve too that its not costly. 
https://github.com/apache/hadoop/pull/1842#discussion_r384066843

Have fixed the wrong retry count getting displayed in log.

Reply:
I see, this is not string concatenation.